File: packages/visionOS/web-spatial/EventEmitter.swift
class EventEmitter {
    private var listeners: [String: [(_ object: Any, _ data: Any) -> Void]] = [:]

    func on(event: String, listener: @escaping (_ object: Any, _ data: Any) -> Void) {
        if listeners[event] == nil {
            listeners[event] = []
        }
        listeners[event]?.append(listener)
    }

    func emit(event: String, data: Any) {
        listeners[event]?.forEach { listener in
            listener(self, data)
        }
    }

    func off(event: String, listener: @escaping (_ object: Any, _ data: Any) -> Void) {
        listeners[event]?.removeAll(where: { $0 as AnyObject === listener as AnyObject })
    }

    func reset() {
        listeners = [:]
    }
}

protocol EventEmitterProtocol {
    var listeners: [String: [(_ object: Any, _ data: Any) -> Void]] { get set }

    mutating func on(event: String, listener: @escaping (_ object: Any, _ data: Any) -> Void)
    func emit(event: String, data: Any)
    mutating func off(event: String, listener: @escaping (_ object: Any, _ data: Any) -> Void)
    mutating func reset()
}

extension EventEmitterProtocol {
    mutating func on(event: String, listener: @escaping (_ object: Any, _ data: Any) -> Void) {
        if listeners[event] == nil {
            listeners[event] = []
        }
        listeners[event]?.append(listener)
    }

    func emit(event: String, data: Any) {
        listeners[event]?.forEach { listener in
            listener(self, data)
        }
    }

    mutating func off(event: String, listener: @escaping (_ object: Any, _ data: Any) -> Void) {
        listeners[event]?.removeAll(where: { $0 as AnyObject === listener as AnyObject })
    }

    mutating func reset() {
        listeners = [:]
    }
}

File: packages/autoTest/src/EventEmitter.ts
import { EventEmitterProtocol } from './types/types'

export class EventEmitter implements EventEmitterProtocol {
  listeners: Record<string, Array<(object: any, data: any) => void>> = {}

  on(event: string, listener: (object: any, data: any) => void): void {
    if (!this.listeners[event]) {
      this.listeners[event] = []
    }
    this.listeners[event].push(listener)
  }

  emit(event: string, data: any): void {
    if (this.listeners[event]) {
      this.listeners[event].forEach(listener => {
        listener(this, data)
      })
    }
  }

  off(event: string, listener: (object: any, data: any) => void): void {
    if (this.listeners[event]) {
      // In TypeScript we remove listeners by index
      const index = this.listeners[event].indexOf(listener)
      if (index !== -1) {
        this.listeners[event].splice(index, 1)
      }
    }
  }

  reset(): void {
    this.listeners = {}
  }
}
